### ISSUE TITLE
Tpetra: tests failing in OpenMP+Cuda and OpenMP+Rocm builds

### ISSUE THREAD

**Issue by romintomasetti** [trilinos/Trilinos#10569]
## Bug Report
@trilinos/tpetra

### Description

When compiling `Trilinos` (*develop branch*) with `OpenMP`+`Cuda` or with `OpenMP`+`Rocm` (`HIP`), we find that *compilation succeeds*, but *some Tpetra tests fail*. 

Specifically, compiling `Trilinos` with:
* `Tpetra`
* `OpenMP`
* `ROCm` (4.5)

works.

We run our tests with
```json
"MPI_EXEC_POST_NUMPROCS_FLAGS"    : {"type" : "STRING", "value" : "--map-by;socket:PE=1;--bind-to;core"},
```

When we set `OMP_NUM_THREADS` to a value **larger than 1**, some `Tpetra` tests with more than one MPI process fail:
```cmake
The following tests FAILED:
25 - TpetraCore_BlockCrsMatrix_MPI_4 (Failed)
55 - TpetraCore_CrsGraph_UnitTests0_MPI_4 (Failed)
56 - TpetraCore_CrsGraph_UnitTests1_MPI_4 (Failed)
57 - TpetraCore_CrsGraph_UnitTests_Swap_MPI_2 (Failed)
58 - TpetraCore_CrsGraph_ReindexColumns_MPI_4 (Failed)
62 - TpetraCore_CrsGraph_getNumDiags_MPI_4 (Failed)
65 - TpetraCore_CrsGraph_StaticImportExport_MPI_4 (Failed)
68 - TpetraCore_CrsGraphTransposer_UnitTests_MPI_4 (Failed)
113 - TpetraCore_Distributor_CreateFromSendsAndRecvs_MPI_4 (Failed)
116 - TpetraCore_FECrsGraph_UnitTests_MPI_4 (Failed)
117 - TpetraCore_FECrsMatrix_UnitTests_MPI_4 (Failed)
124 - TpetraCore_ImportExport_ImportConstructExpert_MPI_4 (Failed)
129 - TpetraCore_UnpackLongRows_MPI_4 (Failed)
138 - TpetraCore_ImportExport2_UnitTests_ISend_MPI_4 (Failed)
195 - TpetraCore_MatrixMatrix_UnitTests_MPI_4 (Failed)
196 - TpetraCore_FECrs_MatrixMatrix_UnitTests_MPI_4 (Failed)
253 - TpetraCore_FEMAssembly_InsertGlobalIndicesFESP_MPI_4 (Failed)
254 - TpetraCore_FEMAssembly_InsertGlobalIndicesFESPKokkos_MPI_4 (Failed)
255 - TpetraCore_FEMAssembly_TotalElementLoopSP_MPI_4 (Failed)
256 - TpetraCore_FEMAssembly_TotalElementLoopSPKokkos_MPI_4 (Failed)
Errors while running CTest
```

The majority of them fail with the following error message (values of host and device use_count may vary):
```c++
Tpetra::Details::WrappedDualView (name = Tpetra::CrsGraph::lclInd; host use_count = 3; device use_count = 2): Cannot access data on host while a device view is alive
```

Running the tests with `OMP_NUM_THREADS=1` works fine though.

We also encountered the same issue when compiling Trilinos with:
* `Tpetra`
* `OpenMP`
* `CUDA` (11.6)

We are therefore in a situation where, if we want to enable a GPU, we must disable `OpenMP.`

Is there something special to do to enable `OpenMP` together with `Cuda` or `ROCm`?

**Comment by csiefer2** [comment]
That is not a configuration we test, so it is unsueprising that it doesn't work.

**Comment by maartenarnst** [comment]
Thanks, @csiefer2!  

In fact, @romintomasetti and I had two purposes in mind:

- On the one hand, yes, we were indeed thinking of asynchronous CPU/GPU assembly/solves.  Here, our question was more exploratory.  We are not working on this actively.
- On the other hand, we were also thinking of day-to-day development of a code in which we want to be able to work with either OpenMP or(and) Cuda, especially as we are working within Docker images (ideally a single Docker image with both).

**Comment by csiefer2** [comment]
Got it.  The single docker for deployment use case makes sense.  

Those tests only fail with OMP_NUM_THREADS > 1 ?

That's a little weird, but I could certainly see if I can reproduce that.

**Comment by romintomasetti** [comment]
@csiefer2 Yes, at that time, I ran the tests enforcing `OMP_NUM_THREADS=1` and they passed...

**Comment by csiefer2** [comment]
I think I can reproduce a bunch of those failures, so it isn't just you.

**Comment by romintomasetti** [comment]
@brian-kelley As discussed last week, this issue seems important to solve. Having a parallel back-end both on the device and on the host should work by default :smile:

**Comment by romintomasetti** [comment]
As far as I'm concerned, I'm good with the fixes brought by @brian-kelley in #11679.

Ok to close this one :wink:
